### ISSUE TITLE
Implement new concept exercise: enum pattern

### ISSUE THREAD

**Issue by andrerfcsantos** [exercism/go#2226]
## Getting Started

If you have not yet contributed to concept exercises before, this task requires some upfront reading to acquire the necessary background knowledge.

Here you can read about what Concept Exercises are and how they are structured:

- [What are Concept Exercises?](https://exercism.org/docs/building/product/concept-exercises)
- [Concept Exercises](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concepts](https://exercism.org/docs/building/tracks/concepts)

Also, be aware of these general guidelines:

- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)

## Goal

Even though Go doesn't have a formal "enum" type, it is possible to have a type definition and then a block of constants of that type and have those constants act as parts of an enumeration. The goal here is to create a new concept exercise that teaches about this pattern.
That includes writing a concept and creating a new exercise.

## Concepts

The following concept needs to be created:

- `enum-pattern`

You can use the `introduction.md` file of the concept also as `introduction.md` file of the exercise. No need to create different content at this point. Additionally, if you want to save some time it is ok to not have an extensive `about.md` at this point. It can also be mainly the `instructions.md` content, maybe with some additions you would like to make.

## Learning Objectives

The student should learn about the following topics and then practice them in the concept exercise:

- given a type definition, create blocks of constants with constants of that type that act as an enum
- show the use of `iota` as a helpful way to give values for the constants easily.

## Prerequisites

- `type-definitions`
- (others might be necessary depending on the concrete exercise)

These prerequisites should be listed in the `config.json` entry for the exercise.
The prerequisites above are a baseline of the prerequisites that might be needed for the exercise.
When implementing the exercise, feel free to add more prerequisites if the exercise needs them.

## Story Idea

It's a good idea for concept exercises to have a story that gives some flair and context to the tasks.
Create a new story for this exercise. The story doesn't need to be highly complex, even a small story goes a long way to make the tasks more interesting. For some inspiration for the story, check the other exercises on the track or the page [Exercism docs: Stories](https://exercism.org/docs/building/tracks/stories) for an overview of stories used across tracks.

## Resources

Some links that might be helpful as a starting point and/or for the links section of the concept:

- [https://www.davidagood.com/go-enum-pattern/](A Pattern For Enums in Go)
- [Working with Enums in Go](https://www.ribice.ba/golang-enums/)
- [Safer Enums in Go](https://threedots.tech/post/safer-enums-in-go/)

## How to proceed

1. **First accept this issue by saying "I'd like to work on this"** (no need to wait for a response, just go ahead).
2. Use this issue to discuss any questions you have, what should be included in the content and whatnot, and to collect more reference material.
3. Create a PR and set "exercism/go" as reviewers. Additionally, you can write in #maintaining-go on Slack that your PR is ready for review. Once you incorporated any critical feedback that the reviewer might give you and the PR is approved, it will be merged by a maintainer.

**Comment by junedev** [comment]
FYI There is a wip exercise in the repo already that covers parts of what is listed above: `savings-account`, concept `constants`.
The exercise would probably need quite some improvement so it might be better to start from scratch. I just wanted to mention it.

Nevertheless, it might be worth thinking about sticking with `constants` as concept name (and adding some non-enum related into about constants) instead of creating `enum` as a concept which does not exist in the language. In the JavaScript track we also discussed this issue of "concepts the language lacks" and decided to better but the things that exist instead in the concept tree rather then the names of the things that don't exist but people might look for. Both ways have pros and cons of course. WDYT?

**Comment by andrerfcsantos** [comment]
Sound good.

I agree with renaming this and making sure we name concepts just after things that are on the language.

If you agree, we can change this issue to be the improvement of that exercise (if any) and we make sure that both constants and this "enum" pattern are well covered.

**Comment by junedev** [comment]
> If you agree, we can change this issue to be the improvement of that exercise (if any) and we make sure that both constants and this "enum" pattern are well covered.

Yes, sounds good.

**Comment by junedev** [comment]
@andrerfcsantos I updated the issue title and the description. Would be great if you could double check it makes sense. Feel free to make more edit as you see fit.

**Comment by andrerfcsantos** [comment]
@junedev Thanks for taking care of this. I took a look and it sounds great!

**Comment by saurami** [comment]
I'd like to work on this

**Comment by junedev** [comment]
@saurami You commented on 2 issues that you would like to work on them. Could you decide for one of them and work on that one first and once it is done pick up the next one?

**Comment by saurami** [comment]
This is going to be my first contribution towards a concept exercise. 

Which one would you recommend ?

**Comment by saurami** [comment]
I'm going to start with the [other PR](https://github.com/exercism/go/issues/1391).

**Comment by junedev** [comment]
Honestly, both tasks have their own difficulties so I couldn't have picked for you. As you made a decision now, I will assign you accordingly.

**Comment by junedev** [comment]
See also https://github.com/exercism/go/issues/1584
